cmd-dev/mt-simulate-memory-idx-lock-contention/main.go

```diff
@@ -0,0 +1,325 @@
+	blockProfileRate     = flag.Int("block-profile-rate", 0, "Sampling rate of block profile, 0 means disabled")
+	mutexProfileFraction = flag.Int("mutex-profile-rate", 0, "Fraction of mutex samples, 0 means disabled")
+	cpuProfile           = flag.Bool("cpu-profile", false, "Enable cpu profile")
+)
+
+func init() {
+	formatter := &logger.TextFormatter{}
+	formatter.TimestampFormat = "2006-01-02 15:04:05.000"
+	log.SetFormatter(formatter)
+	log.SetLevel(log.InfoLevel)
+}
+
+func main() {
+	config, err := globalconf.NewWithOptions(&globalconf.Options{
+		Filename:  "",
+		EnvPrefix: "MT_",
+	})
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "FATAL: configuration file error: %s", err)
+		os.Exit(1)
+	}
+
+	memoryIdxFlags := memory.ConfigSetup()
+
+	flag.Usage = func() {
+		fmt.Println("mt-simulate-memory-idx-lock-contention")
+		fmt.Println()
+		fmt.Println("Simulates index lock congestion")
+		fmt.Println()
+		fmt.Println("Usage:")
+		fmt.Println()
+		fmt.Printf("	mt-simulate-memory-idx-lock-contention [flags] [memory-idx memory-idx flags]\n")
+		fmt.Println()
+		fmt.Println("Flags:")
+		flag.PrintDefaults()
+		fmt.Println("memory-idx flags:")
+		memoryIdxFlags.PrintDefaults()
+	}
+
+	var pos int
+	for pos = 0; pos < len(os.Args); pos++ {
+		if os.Args[pos] == "memory-idx" {
+			break
+		}
+	}
+
+	if pos > 1 {
+		flag.CommandLine.Parse(os.Args[1:pos])
+		memoryIdxFlags.Parse(os.Args[pos+1:])
+	}
+
+	// even if flags have already been parsed, this still allows to use env vars such as MT_MEMORY_IDX_PARTITIONED=true
+	config.ParseAll()
+
+	memory.ConfigProcess()
+
+	runtime.SetBlockProfileRate(*blockProfileRate)
+	runtime.SetMutexProfileFraction(*mutexProfileFraction)
+	if *cpuProfile {
+		filenamePrefix := *profileNamePrefix + ".cpu."
+		f, err := os.Create(findFreeFileName(filenamePrefix))
+		if err != nil {
+			log.Fatalf("Failed to create cpu profile: %s", err.Error())
+		}
+		defer f.Close()
+		if err := pprof.StartCPUProfile(f); err != nil {
+			log.Fatal("could not start cpu profile: ", err)
+		}
+		defer pprof.StopCPUProfile()
+	}
+
+	go func() {
+		log.Println(http.ListenAndServe("localhost:6060", nil))
+	}()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	queryGenerator, err := NewQueryGenerator(ctx, *queriesFile)
+	if err != nil {
+		log.Fatalf("Failed read queries: %s", err.Error())
+	}
+	metricGenerator, err := NewMetricsGenerator(ctx, *seriesFile, *initialIndexSize, *newSeriesPercent)
+	if err != nil {
+		log.Fatalf("Failed read series: %s", err.Error())
+	}
+	testRun := runner.NewTestRun(metricGenerator.Out, queryGenerator.Out, uint32(*addsPerSec), uint32(*addThreads), uint32(*initialIndexSize), uint32(*queriesPerSec), *concQueries)
+	go func() {
+		ticker := time.NewTicker(time.Second * 2)
+		for range ticker.C {
+			runtime.GC()
+		}
+	}()
+
+	testRun.Init(ctx)
+	go testRun.Run()
+
+	// TODO: this looks like a duplicate (see above): confirm we can remove this without affecting results
```

Comment:
I think this duplicate go func can be removed, I can't think of a reason why we'd need it twice so this was probably a mistake